libs/langchain/langchain/schema/runnable.py

```diff
@@ -0,0 +1,683 @@
+            inheritable_tags=config.get("tags"),
+            local_tags=None,
+            inheritable_metadata=config.get("metadata"),
+            local_metadata=None,
+        )
+        # start the root run
+        run_manager = await callback_manager.on_chain_start(
+            dumpd(self), input if isinstance(input, dict) else {"input": input}
+        )
+
+        # invoke the first steps
+        try:
+            for step in [self.first] + self.middle:
+                input = await step.ainvoke(
+                    input,
+                    # mark each step as a child run
+                    _patch_config(config, run_manager.get_child()),
+                )
+        except (KeyboardInterrupt, Exception) as e:
+            await run_manager.on_chain_error(e)
+            raise
+
+        # stream the last step
+        final: Union[Output, None] = None
+        final_supported = True
+        try:
+            async for output in self.last.astream(
+                input,
+                # mark the last step as a child run
+                _patch_config(config, run_manager.get_child()),
+            ):
+                yield output
+                # Accumulate output if possible, otherwise disable accumulation
+                if final_supported:
+                    if final is None:
+                        final = output
+                    else:
+                        try:
+                            final += output  # type: ignore[operator]
+                        except TypeError:
+                            final = None
+                            final_supported = False
+                            pass
+        # finish the root run
+        except (KeyboardInterrupt, Exception) as e:
+            await run_manager.on_chain_error(e)
+            raise
+        else:
+            await run_manager.on_chain_end(
+                final if isinstance(final, dict) else {"output": final}
+            )
+
+
+class RunnableMap(Serializable, Runnable[Input, Dict[str, Any]]):
+    steps: Dict[str, Runnable[Input, Any]]
+
+    @property
+    def lc_serializable(self) -> bool:
+        return True
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def invoke(
+        self, input: Input, config: Optional[RunnableConfig] = None
+    ) -> Dict[str, Any]:
+        from langchain.callbacks.manager import CallbackManager
+
+        # setup callbacks
+        config = config or {}
+        callback_manager = CallbackManager.configure(
+            inheritable_callbacks=config.get("callbacks"),
+            local_callbacks=None,
+            verbose=False,
+            inheritable_tags=config.get("tags"),
+            local_tags=None,
+            inheritable_metadata=config.get("metadata"),
+            local_metadata=None,
+        )
+        # start the root run
+        run_manager = callback_manager.on_chain_start(dumpd(self), {"input": input})
+
+        # gather results from all steps
+        try:
```

Comment:
both sync and async versions are relying on ordering of dict keys/values being consistent on successive invocations -- Might make sense to program defensively here instead

```python
futures_map = {
  key: executor.submit( ...)
   for key, runnable in self.steps.items(
}
}
```